protocol/lib/random_test.go

```diff
@@ -81,3 +82,70 @@ func TestRandomBytesBetween_InvalidInputs(t *testing.T) {
 		RandomBytesBetween([]byte{}, []byte{}, nil)
 	})
 }
+
+func TestWeightedRandomSample(t *testing.T) {
+	source := exprand.NewSource(53)
+
+	freq := make([]float64, 4)
+	for i := 0; i < 1_000_000; i++ {
+		weights := []float64{0.1, 0.2, 0.3, 0.4}
+		sample, err := WeightedRandomSample(weights, 1, source)
+		require.NoError(t, err)
+		freq[sample[0]]++
+	}
+
+	exp := []float64{100_000, 200_000, 300_000, 400_000}
+
+	// Check that this is within statistical expectations.
+	require.Less(
+		t,
+		chi2(freq, exp),
```

Comment:
not sure if we need to do chi2 here, or just compare with the actual expected values. let me know if anyone has a preference